feat(dom_adapter): inject a style with an '!important' rule

DzmitryShylovich opened the pull request:
Fixes https://github.com/angular/angular/issues/12009

Commit:
feat(dom_adapter): allow styles with an '!important' rule

vicb commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -211,14 +211,17 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
+    if (element.style[styleName]) {
+      this.removeStyle(element, styleName);
+    }
+    element.style.cssText += `${styleName}: ${styleValue}`;
```
do you have some ref describing the support across browser ?

DzmitryShylovich replied:
Nope, I didn't find any refs.

Solution mostly based on [answers on SO](http://stackoverflow.com/a/1577204) and https://gist.github.com/tzi/2953155#file-important-js-L5

I've tested it in ie8-11(emulation mode).

vicb replied:
Also shouldn't it be `+= `;${styleName}: ${styleValue}``

vicb replied:
`.style['background-color']` & `.style['backgroundColor']` both work which is not the case with cssText

may be we should restrict for `!important` only ?

DzmitryShylovich replied:
> Also shouldn't it be += ';${styleName}: ${styleValue}'

Browsers always insert `;` between styles correctly so I think it's redundant (tested in ie/edge/chrome).

DzmitryShylovich replied:
> .style['background-color'] & .style['backgroundColor'] both work which is not the case with cssText

I think we should stick with `cssText`. If somebody will complain it will be easier to normilize style name (`backgroundColor` -> `background-color`) instead of parsing style value, substring `important` etc.

DzmitryShylovich replied:
> element.style.setProperty("display", "inline", "important")

also a good solution. hard to decide :)

vicb commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -395,3 +399,12 @@ export function parseCookieValue(cookieStr: string, name: string): string {
 }
+
+const IMPORTANT_RULE: string = '!important';
+function isPriority(styleValue: string): boolean {
```
hasPriority

DzmitryShylovich commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -211,14 +211,18 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
+    if (isPriority(styleValue)) {
+      element.style.setProperty(styleName, removePriority(styleValue), 'important');
+    } else {
+      element.style.setProperty(styleName, styleValue);
```
nope, only  dash-case :(
for camelCase only `element.style.backgroundColor` or `element.style['backgroundColor']`.
so neither `cssText` nor `setProperty` support camelCase. 
We can drop support for camelCase styles or intercept them and convert into dash-case.

vicb replied:
we have to be consistent and fast:
- use special case only for important rule, make sure both case are supported - need to think about vendor prefixes also,
- keep the former code for standard case

We also should think about a way to avoid the `indexOf` for future releases where we can break the API.

DzmitryShylovich commented:
1. Replaced `String.indexOf` with `String.endsWith`. I don't think we can do better without extending existing api to something like `DomAdapter.setStyle(elem: any, styleName: string, styleValue: string, priority: boolean)`
2. Added support for camelCase styles in both cases.
3. Added support for vendor prefixes.

vicb commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -211,14 +212,22 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
+      element.style[styleName] = styleValue;
+    } else {
+      // style.setProperty only supports dash-case style names
+      const dashCaseName = camelCaseToDashCase(styleName);
```
`const dashCaseName = toDashCaseStyleName(styleName);`

where `toDashCaseStyleName` handle both case conversion and vendor prefix

vicb commented on modules/@angular/platform-browser/test/browser/browser_adapter_spec.ts:
```diff
@@ -31,4 +31,94 @@ export function main() {
+
+        it('should properly handle vendor prefixes', () => {
+          dom.setStyle(elem, 'webkitTransform', 'scale(1)');
+          dom.setStyle(elem, 'MozTransform', 'scale(1)');
```
moz

DzmitryShylovich replied:
Nope, tests failed with `mozTransform` (also tested in a browser with no luck) then I found http://stackoverflow.com/a/10962752.
`webkitTransform` and `WebkitTransform` both work in chrome but only `MozTransform` in firefox.

vicb replied:
Thanks for clarification !

vicb commented:
a few nits but LGTM

vicb commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -395,3 +403,23 @@ export function parseCookieValue(cookieStr: string, name: string): string {
+}
+
+const VENDOR_PREFIXES = ['webkit', 'ms', 'o', 'moz'];
+function checkVendorPrefix(input: string): string {
```
please merge into toDashCaseStyleName

vicb commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -395,3 +403,19 @@ export function parseCookieValue(cookieStr: string, name: string): string {
+  return styleValue.replace(IMPORTANT_RULE, '').trim();
+}
+
+const VENDOR_PREFIXES = ['webkit', 'ms', 'o', 'moz'];
```
`Moz` ?

vicb commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -210,15 +211,22 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
+    if (!hasPriority(styleValue)) {
+      element.style[styleName] = styleValue;
+    } else {
+      // style.setProperty only supports dash-case style names
```
group the comments

vicb commented on modules/@angular/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -395,3 +403,19 @@ export function parseCookieValue(cookieStr: string, name: string): string {
+const VENDOR_PREFIXES = ['webkit', 'ms', 'o', 'moz'];
+function toDashCaseStyleName(styleName: string): string {
+  const dashCaseName = camelCaseToDashCase(styleName);
+  const hasVendorPrefix = VENDOR_PREFIXES.some((prefix: string) => dashCaseName.startsWith(prefix));
```
The fact that `Moz` works here is a "friendly" bug but please come with a more robust function. Thanks

DzmitryShylovich replied:
> const dashCaseName = camelCaseToDashCase(styleName);
We are converting camelCase properties to a sash-case (because `element.setPoperty()` method only accepts dash-case).

after that `MozTransition` became `moz-transition` this is why all prefixes are lowercase.

vicb commented:
Could you please rebase this PR
- the code has changed and the DOM adapter is not used any more,
- please add one integration test,
- Do not iterate over [<vendor prefixes>] and create a closure but inline, ie `if/else` calls

DzmitryShylovich commented:
@vicb 
> DOM adapter is not used any more

?

vicb commented:
see https://github.com/angular/angular/commit/d708a8859ca3078a98e371b33ad560096070ad00

DzmitryShylovich commented:
Should I use DomRenderer for Meta service as well?
